### PR TITLE
Hide console toggle button and ignore corresponding keybinding for ipython

### LINKED ISSUE #1531: Button to open console shouldn't exist if viewer launched from IPython
## 🐛 Bug

As noted in https://github.com/napari/napari/issues/787#issuecomment-564829450 the button to open the console should not exist if the viewer is launched from IPython as the console is not supported there (there is no need for it as the IPython terminal functions as the console).

We could consider blocking the console and disabling the button from Jupyter too.

### PR DESCRIPTION
**Pull request by andy-sweet**
# Description
When there is no shell available (which should only happen when starting from ipython or if initialization of QtConsole fails), this PR hides the console toggle button and disables the corresponding toggle behavior of the associated keybinding.

This does not prevent the console from being shown when launching napari from a jupyter notebook, as there seems to be some effort to make that work and I didn't find any issues with it.

One thing I don't like about this PR is that we don't explicitly notify the user that the console is not available when running from ipython. We just silently hide the button and prevent the console from being shown in any other way. Maybe we should make the console button disabled instead of invisible? And maybe we should issue a warning when trying to show it with the keyboard shortcut. I didn't want to unbind the shortcut because that temporarily changes the keyboard shortcut preferences, which seems confusing.

## Type of change

- [x] Bug-fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# References
Close #1531 

# How has this been tested?
- [x] I tested this manually when running from a terminal, from ipython, and from a jupyter notebook.
- [ ] all existing tests pass with my change

I'm not sure how to write tests for this change. Is there a way to make a napari viewer in such a way that we can at least simulate running it from ipython?

## Final checklist:
- [ ] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/docs/dev/guides/translations.html).

### CODE REVIEW

**Comment by tlambert03** [comment]
> We just silently hide the button and prevent the console from being shown in any other way.

I'm actually seeing a visible, clickable button, that disappears when clicked :joy::

https://user-images.githubusercontent.com/1609449/157769847-cc6b7eaa-94a5-48d6-9f78-4924b2af5d4d.mov

I'd be ok with either a grayed out button, or an invisible button... But i wonder whether this could be done more directly [in `qt_viewer_buttons`](https://github.com/napari/napari/blob/d9ac9daf00d720acfb54e3335b1400debd450c69/napari/_qt/widgets/qt_viewer_buttons.py#L119) by just setting that button to hidden or disabled after checking some function (something like the `_ipython_has_eventloop` function in `_qt.qt_event_loop`.  I'm not 100% sure that it will predict the times when we want the console disabled... but I'm ~90% :)

**edit**, sorry, I meant the functions in napari.utils.misc: `in_jupyter` and `in_ipython`

**Comment by andy-sweet** [comment]
> I'm actually seeing a visible, clickable button, that disappears when clicked 😂:

Ooops.

I think that's because in my manual tests the console is always visible initially (not sure why), so it accesses `QtViewer.console` and goes through the code paths I added. But I guess if it's not visible, the `QtViewer.console` has not been accessed yet, and that button is still visible.

> **edit**, sorry, I meant the functions in napari.utils.misc: `in_jupyter` and `in_ipython`

Sounds very promising. Let me try!

**Comment by jni** [comment]
Just in case someone asks, I'm ok with any of the behaviours suggested above, except the magical disappearing button :rofl: Personally, even though it might be surprising to some, I do like the idea of using less space when it is not needed. I also think there's a future where plugins provide buttons and we place them in a button row, which then might be less static than what we have now, and then the button's disappearance under IPython might be less notable?

But, a greyed out button is also totally fine. If the hover text could be changed to "Console in unavailable when using IPython", that would be great from a UX perspective.

**Comment by andy-sweet** [comment]
> I think that's because in my manual tests the console is always visible initially (not sure why),

I'm pretty sure this is because (depending on your preferences/settings), the napari window state can be reinstated in [`Window.show`](https://github.com/napari/napari/blob/6a3e11aa717e7928a0a5a3c7693577729a466ef1/napari/_qt/qt_main_window.py#L1038), which may include a visible `dockConsole`.

I think we should solve that issue because otherwise the only way to hide the console is to hit the tiny-eye button associated with the dock widget (because the larger console button is hidden and the keyboard shortcut is disabled).

My crude solution is to handle this case specifically in `_QtMainWindow._set_window_settings`. That seems to work for me, though I'm slightly worried that it might cause some weird presentation issues for others (e.g. flicker of the dockConsole being shown and hidden).

This PR is pretty small in terms of lines changed, but I don't like how the logic associated with disabling the console when launched from ipython is now spread over three files. One alternative that seems a bit hacky is to do something like the following in `QtViewer.__init__`

```python
...
        if in_ipython():
            self.dockConsole.setVisible = lambda _: None
            self.viewerButtons.consoleButton.setVisible(False)
...
```

but that looks like a bad idea.

Sorry for my ignorance here - I'm pretty unfamiliar with this part of the code base.

**Comment by sofroniewn** [comment]
> This PR is pretty small in terms of lines changed, but I don't like how the logic associated with disabling the console when launched from ipython is now spread over three files.

yeah I don't love this either. in many ways i'd love the console to be treated like any other plugin with no special casing, but the button makes it hard to do that. maybe one day we'll have a mechanism for plugins to add buttons or we'll have a command pallette that makes it super easy to open plugins and we wont need special buttons anymore. I've asked @Carreau to take a look here too

**Comment by Carreau** [comment]
Can we maybe change a "set visible" to maybe grayed-out/disabled ? I think from a UI perspective that would be better. I'm going to think on how we can do somethign better in general, but the console is pretty special.

**Comment by tlambert03** [comment]
> Can we maybe change a "set visible" to maybe grayed-out/disabled ? I think from a UI perspective that would be better. I'm going to think on how we can do somethign better in general, but the console is pretty special.

I like a grayed out button too.  I also agree the console is pretty special.  As it provided deeply integrated functionality before we even had a plugin spec, it's ok with me if we can't completely dogfood this one with the plugin API.

**Comment by psobolewskiPhD** [comment]
As someone who launches napari *a lot* from the terminal and also *a lot* from VS Code Interactive Windows and apparently someone who frequently doesn't pay attention and correlate behavior readily, the console has always been an odd beast. Sometimes I click and it's there and otherwise it's just a black box, blank.
This PR made me *get it*: the console works when I launch from terminal and not when from VS Code! So with that in mind, I would strongly advise to **not** hide the button, but to grey it out (a pretty universal UI for "not available") and, if possible, add a tooltip that says something like *Console not available, use your existing python REPL.* This way a dimwitted person like me can understand what's going on. 

BTW, this reminds me of: https://github.com/napari/napari/issues/3338 and I can add that at present the behavior I posted here https://github.com/napari/napari/issues/3338#issuecomment-913139880 doesn't occur. Launching napari from VS Code does not result in the console inheriting vars, etc. Instead it's a black box.

**Comment by Carreau** [comment]
I pushed a change that switch to enable/disable, and update the tooltip of the action to precise it may be unavailable.

I think though the design needs update of disabled buttons design as they don't look disabled to me, just flat:

Screen Shot 2022-03-16 at 11 27 19

**Comment by andy-sweet** [comment]
> I pushed a change that switch to enable/disable, and update the tooltip of the action to precise it may be unavailable.

Thanks @Carreau! I also pushed a small change with an implementation comment to explain why we need to explicitly hide the console sometimes after restoring the window state. After that, I'm OK with the special handling there.

I still don't like my special handling in `qt_viewer.py` to prevent the keyboard shortcut. I feel like a better way to handle this more generally would be to allow disabling an action in the `ActionManager`. That way we could do something like

```python
if in_python():
    action_manager.disable('napari:toggle_console_visibility')
```

and ideally it would disable the keybinding and button. And maybe even provide a tooltip that is dependent on whether the action is enabled/disabled. But that needs some updates to `ActionManager` that are probably beyond the scope of this PR.

> I think though the design needs update of disabled buttons design as they don't look disabled to me, just flat:

Agreed. I guess this is our first case of disabling this type of button in napari? I notice we have some logic for handling different states of `QPushButton` in `napari/_qt/qt_resources/styles/00_base.qss`. It looks like we are explicitly using the background color from the theme in this case. Does anyone know if it's easy to modulate the intensity (or maybe transparency) of the SVG image declarative (i.e. in a QSS file) or programmatically? We could define separate enabled/disabled SVG images for each button, but then I guess each theme would have to do that which seems like too much to ask.

Overall, I think I'm OK with this PR as it is. But if you think it's worth addressing the disabled style issue before that, I think that's reasonable too.

File: napari/components/_viewer_key_bindings.py [hunk 1]
```diff
@@ -121,6 +121,10 @@ def toggle_selected_visibility(viewer: Viewer):
     viewer.layers.toggle_selected_visibility()
 
 
-@register_viewer_action(trans._("Show/Hide IPython console"))
+@register_viewer_action(
+    trans._(
+        "Show/Hide IPython console (only available when napari started as standalone application)"
+    )
+)
```
Comment (Czaki) [inline napari/components/_viewer_key_bindings.py:128]: maybe part in `()` should be visible only if napari is run from ipython? The end of the message is obsolete for bundle app users and may reduce readability.

**Review by sofroniewn — APPROVED** [review]
LGTM! Maybe @jni can give it a quick spin before merge. Thanks!!

**Pull request merged**